shapes/main.cpp

```diff
@@ -1,96 +1,95 @@
-#include <iostream>
-#include <vector>
 #include <algorithm>
-#include <string>
+#include <functional>
+#include <iostream>
+#include <map>
 #include <memory>
-#include "Shape.hpp"
+#include <string>
+#include <vector>
+
+#include "Circle.hpp"
 #include "Rectangle.hpp"
+#include "Shape.hpp"
 #include "Square.hpp"
-#include "Circle.hpp"
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
+auto sortByArea = [](shared_ptr<Shape> first, shared_ptr<Shape> second) 
 {
-    if(first == nullptr || second == nullptr)
+    if (first == nullptr || second == nullptr){
         return false;
+    }
     return (first->getArea() < second->getArea());
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
+auto perimeterBiggerThan20 = [](shared_ptr<Shape> s) {
+    if (s){
         return (s->getPerimeter() > 20);
+    }
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
-    return false;
+std::function<bool(shared_ptr<Shape>)> areaLessThanX{[x{10}](shared_ptr<Shape> s) {
```

Comment:
auto here as a type is enough. You only need to pass it to findFirstShapeMatchingPredicate as std::function.